content/articles/python-metaclasses-for-metaprogramming/index.md

```diff
@@ -0,0 +1,194 @@
+- `__int__()`: This function enables the initialization of the objects that have been created, and therefore, to call the object, you have to pass it as an argument
+
+The `type()` function may be used directly by the user to construct classes, and there are possibilities of utilizing this function in different ways.
+
+The preceding example shows that the type may be sent in as a single parameter. It may be called with three different arguments by the user. Creating the class will be done by this. Some of the arguments include:
+- Class name
+- Pass the tuple of base classes that class has inherited
+- Class dictionary: Some functions and variables occupy this namespace, which the class will use itself.
+
+```python
+def Town_method(persona):  
+    print("Town class method!")
+class Base:  
+    def enjoyerfunction(persona):  
+        print("Inherited method!")
+Town = type('Town', (Base, ), dict(b = "John Doe", enjoyer_method = Town_method))
+print("The Sort of Town class: ", type(Town))
+Town_object = Town()  
+print(" The Sort of Town_object: ", type(Town_object)) 
+Town_object.enjoyerfunction() 
+Town_object.enjoyer_method() 
+print(Town_object.b) 
+```
+
+#### Output
+
+```python
+The Sort of Town class:  <class 'type'>
+The Sort of Town_object:  <class '__main__.Town'>
+Inherited method!
+Town class method!
+John Doe
+```
+
+We will create the basic class then use the `type()` method to generate the city class dynamically. We will put the City type there and construct a City class instance. The inherited method will be called next. The City class method is what we will be using here. Finally, we will see the variable.
+
+### Using a metaclass to fix an issue
+Some of the challenges users face may be addressed with the help of metaclasses and decorators. However, a metaclass is the only approach to address specific issues. For example, when debugging, the body of class functions should always be executed before the fully qualified name is displayed.
+
+```python
+from functools import wraps  
+def debugg(funct):  
+    '''The debugging decorator for passing functions'''  
+    @wraps(funct)  
+    def wrapperr(*args, **kwargs):  
+        print("This function's full name is:", funct.__qualname__)  
+        return funct(*args, **kwargs)  
+    return wrapperr  
+def debug_methods(clas):  
+    '''class decorator make use of debug decorator  
+       for debuging the class functions '''  
+    for key, value in vars(clas).items():  
+        if callable(value):  
+            setattr(clas, key, debugg(value))  
+    return clas 
+@debugmethods  
+class Calc:  
+    def add(own, x, y):  
+        return x+y  
+    def mul(own, x, y):  
+        return x*y  
+    def div(own, x, y):  
+        return x/y 
+       
+enjoyer_cal = Calc()  
+print(enjoyer_cal.add(6, 4))  
+print(enjoyer_cal.mul(8, 9))  
+print(enjoyer_cal.div(30, 5)) 
+```

```

Comment:
Explain the code snippet here well. Tell the user what it does and how it does it.